src/mips/psyqo/src/advancedpad.cpp

```diff
@@ -0,0 +1,305 @@
+/*
+
+MIT License
+
+Copyright (c) 2023 PCSX-Redux authors
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+*/
+
+#include "psyqo/advancedpad.hh"
+
+#include "common/syscalls/syscalls.h"
+#include "psyqo/hardware/cpu.hh"
+#include "psyqo/hardware/sio.hh"
+#include "psyqo/kernel.hh"
+
+using namespace psyqo::Hardware;
+
+void psyqo::AdvancedPad::initialize() {
+    // Stop the kernel from processing pad(and card) events
+    syscall_stopPad();
+
+    for (int i = 0; i < 8; i++) {
+        __builtin_memset(m_padData[i], 0xff, sizeof(m_padData[i]));
+    }
+
+    // Init Pad
+    syscall_memset(m_padData[0], 0xff, sizeof(m_padData[0]));
+
+    SIO::Ctrl = SIO::Control::CTRL_IR;
+    SIO::Baud = 0x88;  // 250kHz
+    SIO::Mode = 0xd;   // MUL1, 8bit, no parity, normal polarity
+    SIO::Ctrl = 0;
+
+    using namespace timer_literals;
+
+    Kernel::Internal::addOnFrame([this]() {
+        readPad();
```

Comment:
So one detail here is that the callbacks put on `addOnFrame` are run at the beginning of every drawn frame. If the software runs at 10FPS, then it'll be called 10 times per second, which may make it lose events, potentially. Same as with the SimplePad, really. But we can change this later.